<!-- Simulate forklift using turtlebot to test approach control -->
<launch>
    <!-- Global Arguments and Parameters -->
    <arg name="debug" default="false"/>
    <param name="/roll/radius" value="0.20"/>

    <!-- Start up the joystick -->
    <arg name="joy_config" default="logitech" />
    <arg name="joy_dev" default="/dev/input/js0" />
    <arg name="config_filepath" default="$(find teleop_twist_joy)/config/$(arg joy_config).config.yaml" />

    <node pkg="joy" type="joy_node" name="joy_node">
        <param name="dev" value="$(arg joy_dev)" />
        <param name="deadzone" value="0.3" />
        <param name="autorepeat_rate" value="20" />
    </node>

    <!-- Begin turtlebot -->
    <!-- Simulated Turtlebot -->
    <include file="$(find turtlebot_gazebo)/launch/turtlebot_world.launch">
        <arg name="world_file" value="/opt/ros/kinetic/share/turtlebot_gazebo/worlds/empty.world"/>
        <arg name="gui" value="false"/>
    </include>
    <!-- Read Turtlebot -->
    <!--
    <include file="$(find turtlebot_bringup)/launch/minimal.launch"/>
    -->

    <!-- Start up the lidar -->
    <include file="$(find ouster_ros)/launch/lidar2.launch"/>

    <!-- Start the cylinder detection -->
    <node pkg="grasping" name="cylinder_detection" type="cylinder_detection">
        <param name="debug" value="$(arg debug)"/>
        <param name="target_frame" value="odom"/>
        <param name="target_x" value="0.0"/>
        <param name="target_y" value="1.0"/>
    </node>

    <!-- Run approach controller -->
    <node pkg="grasping" name="approach_controller" type="approach_controller">
        <param name="target_x" value="2.0"/>
        <param name="target_y" value="0.0"/>
        <param name="grasp_angle" value="0.7854"/>
        <param name="angle_tolerance" value="0.01"/>
        <param name="K_angle" value="1.0"/>
        <param name="K_linear" value="1.0"/>
        <param name="max_velocity" value="1.0"/>
        <param name="grasp_plate_offset_x" value="0.36"/>
        <param name="grasp_plate_offset_y" value="-0.18"/>
        <param name="approach_offset" value="2.0"/>
        <param name="approach_tolerance" value="0.01"/>
        <param name="backout_distance" value="1.0"/>
    </node>

    <!-- Convert velocity and angle setpoints into a Twist command -->
    <node pkg="sensors" name="velocity_conversion" type="velocity_conversion">
        <param name="use_covariance" value="false"/>
        <remap from="/velocity_node/velocity" to="/velocity_node/velocity_setpoint"/>
        <remap from="/steering_node/filtered_angle" to="/steering_node/angle_setpoint"/>
        <remap from="/velocity_conversion/twist" to="/mobile_base/commands/velocity"/>
    </node>
</launch>
